doc/man/grephistory.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GREPHISTORY 1"
.TH GREPHISTORY 1 "2016-10-07" "INN 2.6.1" "InterNetNews Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
grephistory \- Query the INN history database
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBgrephistory\fR [\fB\-eilnqsv\fR] [\fB\-f\fR \fIdb\fR] [\fImessage-id\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBgrephistory\fR queries the \s-1INN\s0 history database for information about the
specified message-ID.  If no flags are given, the program prints the
storage \s-1API\s0 token of the corresponding article, or \f(CW\*(C`/dev/null\*(C'\fR if the
article is listed in the history database but not stored on the server.
If the message-ID cannot be found in the database, \fBgrephistory\fR will
print \f(CW\*(C`grephistory: not found\*(C'\fR and exit with a non-zero status.
.PP
Be sure to escape any special characters in the message \s-1ID\s0 from the shell.
Single quotes are recommended for this purpose since many message-IDs
contain dollar signs.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-e\fR" 4
.IX Item "-e"
Only print the storage token if the article is stored on the system.  (In
other words, suppress the \f(CW\*(C`/dev/null\*(C'\fR or \f(CW\*(C`not found\*(C'\fR output for missing
or remembered articles.)
.IP "\fB\-f\fR \fIdb\fR" 4
.IX Item "-f db"
Query the history database \fIdb\fR rather than the default history database.
.IP "\fB\-i\fR" 4
.IX Item "-i"
Rather than expecting a message-ID on the command line, \fBgrephistory\fR
will read a list of message-IDs on standard input, one per line.  Leading
and trailing whitespace is ignored, as are any malformed lines.  It will
print on standard output those message-IDs which are not found in the
history database.  This is used when processing \f(CW\*(C`ihave\*(C'\fR control messages.
.IP "\fB\-l\fR" 4
.IX Item "-l"
Display the entire line from the history database, rather than just the
storage \s-1API\s0 token.  If the message-ID is present in the history database
but has no storage \s-1API\s0 token, \fBgrephistory\fR does not print anything.
.IP "\fB\-n\fR" 4
.IX Item "-n"
If the message-ID is present in the history database but has no storage
\&\s-1API\s0 token, print \f(CW\*(C`/dev/null\*(C'\fR and exit successfully.  This can happen if
an article has been cancelled or expired, but history information has
still been retained.  This is the default behavior.
.IP "\fB\-q\fR" 4
.IX Item "-q"
Don't print any message, but still exit with the appropriate status.
.IP "\fB\-s\fR" 4
.IX Item "-s"
Rather than expecting a message-ID on the command line, \fBgrephistory\fR
will read a list of message-IDs on standard input, one per line.  Leading
and trailing whitespace is ignored, as are any malformed lines.  It will
print on standard output the storage \s-1API\s0 tokens for any articles that are
still available, one per line.  This flag is used when processing
\&\f(CW\*(C`sendme\*(C'\fR control messages.
.IP "\fB\-v\fR" 4
.IX Item "-v"
Print out the hash of the message-ID for diagnostic purposes, as well as
any other requested information.  This flag is not useful with \fB\-i\fR or
\&\fB\-s\fR.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
In case the requested article is not listed in the history database:
.PP
.Vb 4
\&    % grephistory \*(Aq<unknown.article@news.example.com>\*(Aq
\&    grephistory: not found
\&    % grephistory \-v \*(Aq<unknown.article@news.example.com>\*(Aq
\&    grephistory: not found (hash is 501C66C22932BA91131186D7218201EB)
.Ve
.PP
In case the requested article is listed in the history database
but not stored on the server:
.PP
.Vb 2
\&    % grephistory \*(Aq<remembered.article@news.example.com>\*(Aq
\&    /dev/null
.Ve
.PP
In case the requested article is stored on the server:
.PP
.Vb 4
\&    % grephistory \*(Aq<87fxeaay1z.fsf@windlord.stanford.edu>\*(Aq
\&    @02014A2DD6231FCC00000000000000000000@
\&    % grephistory \-l \*(Aq<87fxeaay1z.fsf@windlord.stanford.edu>\*(Aq
\&    [B6DDF69376E3CC199246CEC949B3ACAC]  1244517923~\-~1244517912  @02014A2DD6231FCC00000000000000000000@
.Ve
.PP
With \fBsm\fR, we can retrieve the article, and get its posting date:
.PP
.Vb 2
\&    % grephistory \*(Aq<87fxeaay1z.fsf@windlord.stanford.edu>\*(Aq | sm | grep Date
\&    Date: Mon, 08 Jun 2009 20:25:12 \-0700
\&
\&    % convdate \-dc 1244517912
\&    Tue, 9 Jun 2009 03:25:12 \-0000 (UTC)
.Ve
.PP
It matches the number recorded in history as for its posting date.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Rich \f(CW$alz\fR <rsalz@uunet.uu.net> for InterNetNews.  Rewritten in
\&\s-1POD\s0 by Russ Allbery <eagle@eyrie.org>.
.PP
\&\f(CW$Id:\fR grephistory.pod 9971 2015\-12\-10 20:30:10Z iulius $
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIhistory\fR\|(5), \fIinn.conf\fR\|(5), \fIsm\fR\|(1).


NEWS
Changes in 2.6.1

    * nnrpd now uses -0000 as the time zone for Date: and Injection-Date:
      header fields it generates.  It was previously using +0000, wrongly
      systematically indicating a local time zone at Universal Time when
      *localtime* is set to false (which is the default) in readers.conf. 
      The +0000 time zone will now be used only if *localtime* is set to
      true and UTC is really the local time zone of the server.

    * Julien Elie has implemented in nnrpd the new COMPRESS command
      described in draft-murchison-nntp-compress that extends the NNTP
      protocol to allow a connection to be effectively and efficiently
      compressed.  News clients that also support that extension will be
      able to benefit from that bandwidth optimization and improvement in
      speed.  Moreover, using COMPRESS is more secure than TLS-level
      compression, as far as authentication credentials are concerned.

    * The default value for the *tlscompression* parameter in inn.conf has
      changed.  TLS-level compression is now disabled by default, to comply
      with the best current practices for a secure use of TLS in application
      protocols like NNTP.  Using the new COMPRESS command is recommended.

    * The *tlscompression* parameter in inn.conf now also permits to disable
      TLS-level compression with OpenSSL 0.9.8.  It previously had an effect
      only when OpenSSL 1.0.0 or later was used.

    * rnews no longer segfaults at startup when started setuid news.  Thanks
      to Marcus Jodorf for the bug report.

    * Fixed slow nnrpd responses for a few NNTP commands.  The TCP_NODELAY
      option was unconditionally set whereas only BSD/OS systems needed it. 
      Thanks to Christian Mock for having discovered that.

    * Articles containing a Received: or a Posted: header field are no
      longer rejected by nnrpd at injection time.

    * Articles containing control characters or whitespace-only content
      lines in their headers are now rejected by nnrpd at injection time.

    * OpenSSL 1.1.0 support has been added to INN.

    * When an encryption layer is negotiated during a successful use of the
      STARTTLS command, or after a successful authentication using a SASL
      mechanism that negotiates an encryption layer, nnrpd now updates the
      permissions of the news client according to the new secure state of
      his connection (that is to say auth blocks in readers.conf using the
      *require_ssl* parameter are taken into account).  Previously, only
      connections on a dedicated port (usually 563) were taking benefit from
      that parameter.  Thanks to Steve Crook for the bug report.

    * When a data integrity layer was negotiated during a successful SASL
      authentication, nnrpd was wrongly reseting any knowledge obtained from
      the client, such as the current newsgroup and article number.  This
      behaviour now applies only when an encryption layer is negotiated.

    * nntpsend now correctly waits until all of the child innxmit processes
      exit before it does.  It was causing nntpsend to fail to work properly
      on systems that use systemd, because when it exits prematurely,
      systemd kills all of the processes it launched, including the innxmit
      processes.  Thanks to Jonathan Kamens for the patch.

    * Update from GNU Libtool 2.4.2 to 2.4.6.

    * Other minor bug fixes and documentation improvements.

Upgrading from 2.5 to 2.6

    The following changes require your full attention because a manual
    intervention may be needed:

    * The name and location of the pullnews configuration file have changed.
      It is now pullnews.marks, located in *pathdb* when pullnews is run as
      the news user, or otherwise in the running user's home directory. 
      This file was previously stored in .pullnews in the running user's
      home directory (even for the news user).  If you use pullnews, you
      need to manually move and rename the configuration file; otherwise, it
      will no longer work.  Note that the -c flag passed to pullnews allows
      to specify another configuration file, if need be.

    * The default location of the mailpost database directory has changed
      from *pathtmp* to *pathdb*.  If you use mailpost without an explicitly
      specified database directory (using the -b flag), then you should
      manually move your current database files mailpost-msgid.dir and
      mailpost-msgid.pag from *pathtmp* to *pathdb*.

    * If you have been using TLS/SSL with nnrpd before, be aware that the
      default value of a few inn.conf parameters have changed:  the server
      now decides the preferred cipher (instead of the client), and only TLS
      protocols are allowed (using the flawed SSLv2 and SSLv3 protocols is
      now disabled).  If you want to change these settings, the respective
      *tlspreferserverciphers* and *tlsprotocols* parameters can be tuned to
      your needs.

    * The --with-kerberos configure flag used to add Kerberos v5 support has
      been renamed to --with-krb5.

    * The --with-berkeleydb configure flag used to add Berkeley DB support
      has been renamed to --with-bdb.

    * The --enable-ipv6 configure flag no longer exists.  IPv6 is now
      unconditionally enabled, if available.

    * $HOME is no longer exported as an environment variable by
      innshellvars, innshellvars.tcl and the Perl module "INN::Config".  It
      was previously overriding the default user home directory with
      *pathnews*.  If you use these scripts in your own scripts, you will
      have to take care of that change.

    * Owing to the implementation of RFC 4643 (AUTHINFO USER/PASS) in innd,
      if remote peers have to authenticate in order to feed articles, they
      now have to send a username (which was previously wrongly optional),
      before sending their password.  The mandatory username, though

expire/grephistory.c
/*  $Id: grephistory.c 8578 2009-08-18 20:11:22Z iulius $
**
**  Get data from history database.
*/

#include "clibrary.h"
#include <syslog.h>  
#include <sys/stat.h>

#include "inn/history.h"
#include "inn/innconf.h"
#include "inn/messages.h"
#include "inn/libinn.h"
#include "inn/paths.h"
#include "inn/storage.h"

/*
**  Read stdin for list of Message-ID's, output list of ones we
**  don't have.  Or, output list of files for ones we DO have.
*/
static void
IhaveSendme(struct history *h, char What)
{
    char		*p;
    char		*q;
    char		buff[BUFSIZ];

    while (fgets(buff, sizeof buff, stdin) != NULL) {
	time_t arrived, posted, expires;
	TOKEN token;

	for (p = buff; ISWHITE(*p); p++)
	    ;
	if (*p != '<')
	    continue;
	for (q = p; *q && *q != '>' && !ISWHITE(*q); q++)
	    ;
	if (*q != '>')
	    continue;
	*++q = '\0';

	if (!HIScheck(h, p)) {
	    if (What == 'i')
		printf("%s\n", p);
	    continue;
	}

	/* Ihave -- say if we want it, and continue. */
	if (What == 'i') {
	    continue;
	}

	if (HISlookup(h, p, &arrived, &posted, &expires, &token))
	    printf("%s\n", TokenToText(token));
    }
}


/*
**  Print a usage message and exit.
*/
static void
Usage(void)
{
    fprintf(stderr, "Usage: grephistory [flags] MessageID\n");
    exit(1);
}


int
main(int ac, char *av[])
{
    int			i;
    const char		*History;
    char                *key;
    char		What;
    struct history	*history;
    time_t arrived, posted, expires;
    TOKEN token;
    unsigned int	Verbosity = 0;

    /* First thing, set up logging and our identity. */
    openlog("grephistory", L_OPENLOG_FLAGS | LOG_PID, LOG_INN_PROG);
    message_program_name = "grephistory";

    /* Set defaults. */
    if (!innconf_read(NULL))
        exit(1);

    History = concatpath(innconf->pathdb, INN_PATH_HISTORY);

    What = '?';

    /* Parse JCL. */
    while ((i = getopt(ac, av, "vf:eilnqs")) != EOF)
	switch (i) {
	default:
	    Usage();
	    /* NOTREACHED */
	case 'v':
	    Verbosity++;
	    break;
	case 'f':
	    History = optarg;
	    break;
	case 'e':
	case 'i':
	case 'l':
	case 'n':
	case 'q':
	case 's':
	    if (What != '?') {
                die("only one [eilnqs] flag allowed");
	    }
	    What = (char)i;
	    break;
	}
    ac -= optind;
    av += optind;

    history = HISopen(History, innconf->hismethod, HIS_RDONLY);
    if (history == NULL)
        die("cannot open history");

    /* Set operating mode. */
    switch (What) {
    case '?':
	What = 'n';
	break;
    case 'i':
    case 's':
	IhaveSendme(history, What);
	HISclose(history);
	exit(0);
	/* NOTREACHED */
    }

    /* All modes other than -i and -l want a message-ID. */
    if (ac != 1)
	Usage();

    key = av[0];
    if (*key == '[') {
        warn("accessing history by hash isn't supported");
	HISclose(history);
	exit(1);
    } else {
        /* Add optional braces if not already present. */
	if (*key != '<')
            key = concat("<", key, ">", (char *) 0);
    }

    if (!HIScheck(history, key)) {
	if (What == 'n') {
	    if (Verbosity > 0)
		die("not found (hash is %s)", HashToText(HashMessageID(key)));
	    else
		die("not found");
	}
    }
    else if (What != 'q') {
	if (HISlookup(history, key, &arrived, &posted, &expires, &token)) {
	    if (What == 'l') {
                if (expires <= 0) {
                    printf("[%s]\t%ld~-~%ld\t%s\n",
                           HashToText(HashMessageID(key)),
                           (long)arrived, (long)posted, TokenToText(token));
                } else {
                    printf("[%s]\t%ld~%ld~%ld\t%s\n",
                           HashToText(HashMessageID(key)),
                           (long)arrived, (long)expires, (long)posted,
                           TokenToText(token));
                }
	    }
	    else {
		if (Verbosity > 0)
		    printf("%s (hash is %s)\n", TokenToText(token),
			    HashToText(HashMessageID(key)));
		else
		    printf("%s\n", TokenToText(token));
	    }
	}
        else if (What == 'n') {
            if (Verbosity > 0)
                printf("/dev/null (hash is %s)\n", HashToText(HashMessageID(key)));
            else
                printf("/dev/null\n");
        }
    }
    HISclose(history);
    return 0;
}
